File: clients/android/app/src/main/java/com/rookery/rook/net/RookApi.kt
// Mirrors clients/RookKit/Sources/RookKit/Net/RookAPI.swift
package com.rookery.rook.net

import com.rookery.rook.model.AgentDefinition
import com.rookery.rook.model.AgentSessionSummary
import com.rookery.rook.model.CandidateEnvironmentRecord
import com.rookery.rook.model.EnvironmentCandidate
import com.rookery.rook.model.EnvironmentListItem
import com.rookery.rook.model.EnvironmentPreview
import com.rookery.rook.model.IdentifyAvailableRequest
import com.rookery.rook.model.get
import com.rookery.rook.model.stringValue
import com.rookery.rook.model.boolValue
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import kotlinx.serialization.Serializable
import kotlinx.serialization.builtins.ListSerializer
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonArray
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.JsonNull
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.buildJsonObject
import kotlinx.serialization.json.decodeFromJsonElement
import kotlinx.serialization.json.encodeToJsonElement
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.put
import okhttp3.HttpUrl
import okhttp3.HttpUrl.Companion.toHttpUrl
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import java.util.concurrent.TimeUnit

class RookApiException(message: String) : Exception(message)

/** Distinguishes 401 from other failures so the UI can show an UNAUTHORIZED state. */
sealed class RookHealthResult {
    data object Ok : RookHealthResult()
    data object Unauthorized : RookHealthResult()
    data class HttpStatus(val code: Int) : RookHealthResult()
    data class TransportError(val message: String) : RookHealthResult()
}

/** REST control plane for the Rook server. */
class RookApi(
    val baseUrl: String = "http://127.0.0.1:3000",
    authToken: String = ""
) {
    private val base: HttpUrl = baseUrl.toHttpUrl()
    private val client = OkHttpClient()
    private val json = Json { ignoreUnknownKeys = true }

    // Trimmed, empty-as-absent — matches RookAPI.swift's authToken normalization.
    private val token: String = authToken.trim()

    val webSocketUrl: String
        get() {
            val httpForm = base.newBuilder().encodedPath("/api/ws").build().toString()
            return if (base.scheme == "https") httpForm.replaceFirst("https://", "wss://")
            else httpForm.replaceFirst("http://", "ws://")
        }

    val webAppUrl: String get() = baseUrl

    suspend fun healthResult(timeoutMs: Long = 1500): RookHealthResult = withContext(Dispatchers.IO) {
        try {
            val timedClient = client.newBuilder().callTimeout(timeoutMs, TimeUnit.MILLISECONDS).build()
            val request = authorized(Request.Builder().url(requestUrl("api/health"))).build()
            timedClient.newCall(request).execute().use { response ->
                when (response.code) {
                    200 -> {
                        val body = runCatching { json.parseToJsonElement(response.body?.string().orEmpty()) }.getOrNull()
                        if (body?.get("ok")?.boolValue == true) RookHealthResult.Ok
                        else RookHealthResult.TransportError("Malformed health response")
                    }
                    401 -> RookHealthResult.Unauthorized
                    else -> RookHealthResult.HttpStatus(response.code)
                }
            }
        } catch (e: Exception) {
            RookHealthResult.TransportError(e.message ?: "Transport error")
        }
    }

    suspend fun health(timeoutMs: Long = 1500): Boolean = healthResult(timeoutMs) is RookHealthResult.Ok

    suspend fun agents(): List<AgentDefinition> {
        @Serializable data class RuntimeResponse(val runtimes: List<AgentDefinition>)
        val body = getJson("api/agent_runtimes")
        return json.decodeFromJsonElement(RuntimeResponse.serializer(), body).runtimes
    }

    suspend fun environmentPreview(environmentId: String): EnvironmentPreview {
        val body = getJson("api/environments/preview", mapOf("environmentId" to environmentId))
        return json.decodeFromJsonElement(EnvironmentPreview.serializer(), body)
    }

    suspend fun registerEnvironment(candidate: CandidateEnvironmentRecord) {
        val payload = json.encodeToJsonElement(CandidateEnvironmentRecord.serializer(), candidate).jsonObject
        postJson("api/environments/register", payload)
    }

    /**
     * Ask the server which `location:` environments are likely available at the given
     * location. Identification only — does not register/enter anything.
     */
    suspend fun identifyAvailableEnvironments(request: IdentifyAvailableRequest): List<EnvironmentCandidate> {
        @Serializable data class IdentifyResponse(val candidates: List<EnvironmentCandidate>)
        val payload = json.encodeToJsonElement(IdentifyAvailableRequest.serializer(), request).jsonObject
        val body = postJson("api/environments/identify", payload)
        return json.decodeFromJsonElement(IdentifyResponse.serializer(), body).candidates
    }

    /**
     * Committing variant: identify, then register/auto-enter the dwell set into the
     * current session/runtime flow. This is what the arrival path (classifier Stationary) calls.
     */
    suspend fun registerLocation(request: IdentifyAvailableRequest): List<EnvironmentCandidate> {
        @Serializable data class IdentifyResponse(val candidates: List<EnvironmentCandidate>)
        val payload = json.encodeToJsonElement(IdentifyAvailableRequest.serializer(), request).jsonObject
        val body = postJson("api/environments/register-location", payload)
        return json.decodeFromJsonElement(IdentifyResponse.serializer(), body).candidates
    }

    suspend fun decideEnvironment(environmentId: String, bundleHash: String, decision: String) {
        postJson(
            "api/environments/decision",
            buildJsonObject {
                put("environmentId", environmentId)
                put("bundleHash", bundleHash)
                put("decision", decision)
            }
        )
    }

    suspend fun enterEnvironment(sessionId: String, environmentId: String): List<String> =
        updateSessionEnvironments(sessionId, listOf(environmentId), emptyList())

    suspend fun exitEnvironment(sessionId: String, environmentId: String): List<String> =
        updateSessionEnvironments(sessionId, emptyList(), listOf(environmentId))

    suspend fun updateSessionEnvironments(sessionId: String, enterEnvironmentIds: List<String>, leaveEnvironmentIds: List<String>): List<String> {
        @Serializable data class EnterResponse(val ok: Boolean = false, val entered: List<String> = emptyList())
        val body = postJson(
            "api/session/environments",
            buildJsonObject {
                put("sessionId", sessionId)
                put("enterEnvironmentIds", json.encodeToJsonElement(ListSerializer(String.serializer()), enterEnvironmentIds))
                put("leaveEnvironmentIds", json.encodeToJsonElement(ListSerializer(String.serializer()), leaveEnvironmentIds))
            }
        )
        return json.decodeFromJsonElement(EnterResponse.serializer(), body).entered
    }

    // GET returns a bare array (mirrors RookAPI.swift environmentList).
    suspend fun environmentList(sessionId: String): List<EnvironmentListItem> {
        val body = getJson("api/environments/list", mapOf("sessionId" to sessionId))
        return json.decodeFromJsonElement(ListSerializer(EnvironmentListItem.serializer()), body)
    }

    // MARK: - Transport helpers

    private fun requestUrl(path: String, query: Map<String, String> = emptyMap()): HttpUrl {
        val builder = base.newBuilder().addPathSegments(path)
        query.forEach { (key, value) -> builder.addQueryParameter(key, value) }
        return builder.build()
    }

    // Single chokepoint for the auth header — mirrors RookAPI.swift authorizedRequest.
    private fun authorized(builder: Request.Builder): Request.Builder {
        if (token.isNotEmpty()) builder.header("Authorization", "Bearer $token")
        return builder
    }

    private suspend fun getJson(path: String, query: Map<String, String> = emptyMap()): JsonElement {
        val request = authorized(Request.Builder().url(requestUrl(path, query))).build()
        return execute(request)
    }

    private suspend fun postJson(path: String, payload: JsonElement): JsonElement {
        val body = json.encodeToString(JsonElement.serializer(), payload).toRequestBody("application/json".toMediaType())
        val request = authorized(Request.Builder().url(requestUrl(path)).post(body)).build()
        return execute(request)
    }

    private suspend fun execute(request: Request): JsonElement = withContext(Dispatchers.IO) {
        client.newCall(request).execute().use { response ->
            val text = response.body?.string().orEmpty()
            if (response.code >= 400) {
                val errorBody = runCatching { json.parseToJsonElement(text) }.getOrNull()
                val message = errorBody?.get("error")?.stringValue ?: "Server error (${response.code})"
                throw RookApiException(message)
            }
            if (text.isEmpty()) JsonNull else json.parseToJsonElement(text)
        }
    }
}

File: clients/RookKit/Sources/RookKit/Net/RookAPI.swift
import Foundation

public struct RookAPIError: LocalizedError {
    public let message: String

    public init(message: String) {
        self.message = message
    }

    public var errorDescription: String? { message }
}

public enum RookHealthResult: Equatable {
    case ok
    case unauthorized
    case httpStatus(Int)
    case transportError(String)
}

/// REST control plane for the Rook server.
public struct RookAPI {
    public let baseURL: URL
    public let authToken: String?

    public init(baseURL: URL = URL(string: "http://127.0.0.1:3000")!, authToken: String? = nil) {
        self.baseURL = baseURL
        let trimmed = authToken?.trimmingCharacters(in: .whitespacesAndNewlines)
        self.authToken = (trimmed?.isEmpty == false) ? trimmed : nil
    }

    public var webSocketURL: URL {
        var components = URLComponents(url: baseURL, resolvingAgainstBaseURL: false)!
        components.scheme = baseURL.scheme == "https" ? "wss" : "ws"
        components.path = "/api/ws"
        return components.url!
    }

    public var webAppURL: URL { baseURL }

    public func webSocketRequest(sessionId: String) -> URLRequest {
        var components = URLComponents(url: webSocketURL, resolvingAgainstBaseURL: false)!
        components.queryItems = [URLQueryItem(name: "sessionId", value: sessionId)]
        var request = authorizedRequest(url: components.url!)
        request.timeoutInterval = 30
        return request
    }

    public func healthResult(timeout: TimeInterval = 1.5) async -> RookHealthResult {
        var request = authorizedRequest(url: baseURL.appending(path: "api/health"))
        request.timeoutInterval = timeout
        do {
            let (data, response) = try await URLSession.shared.data(for: request)
            guard let http = response as? HTTPURLResponse else {
                return .transportError("Non-HTTP response")
            }
            switch http.statusCode {
            case 200:
                if let body = try? JSONDecoder().decode(JSONValue.self, from: data), body["ok"]?.boolValue == true {
                    return .ok
                }
                return .transportError("Malformed health response")
            case 401:
                return .unauthorized
            default:
                return .httpStatus(http.statusCode)
            }
        } catch {
            return .transportError(error.localizedDescription)
        }
    }

    public func healthStatus(timeout: TimeInterval = 1.5) async -> Int? {
        switch await healthResult(timeout: timeout) {
        case .ok: return 200
        case .unauthorized: return 401
        case .httpStatus(let code): return code
        case .transportError: return nil
        }
    }

    public func health(timeout: TimeInterval = 1.5) async -> Bool {
        if case .ok = await healthResult(timeout: timeout) {
            return true
        }
        return false
    }

    public func agents() async throws -> [AgentDefinition] {
        struct AgentsResponse: Codable {
            let agents: [AgentDefinition]
        }
        let body: AgentsResponse = try await get(path: "api/agents", query: [:])
        return body.agents
    }

    public func sessions(agent: String) async throws -> [AgentSessionSummary] {
        let body = try await getJSON(path: "api/agent/sessions", query: ["agent": agent])
        guard case .array(let items)? = body["sessions"] else {
            return []
        }
        return items.map(AgentSessionSummary.init(raw:))
    }

    public func recentSession() async throws -> AgentSessionSummary? {
        let body = try await getJSON(path: "api/agent/session/recent", query: [:])
        guard let session = body["session"], session != .null else {
            return nil
        }
        return AgentSessionSummary(raw: session)
    }

    public func startSession(agent: String, sessionName: String?) async throws -> AgentSessionSummary {
        var payload: [String: JSONValue] = ["agent": .string(agent)]
        if let sessionName, !sessionName.isEmpty {
            payload["sessionName"] = .string(sessionName)
        }
        return try await start(payload: payload)
    }

    public func resumeSession(_ session: AgentSessionSummary) async throws -> AgentSessionSummary {
        let payload: [String: JSONValue] = [
            "agent": .string(session.agent),
            "session": session.raw,
        ]
        return try await start(payload: payload)
    }

    private func start(payload: [String: JSONValue]) async throws -> AgentSessionSummary {
        let body = try await postJSON(path: "api/agent/start", payload: .object(payload))
        guard let session = body["session"], session != .null else {
            throw RookAPIError(message: "Server returned no session")
        }
        return AgentSessionSummary(raw: session)
    }

    public func environmentPreview(environmentId: String) async throws -> EnvironmentPreview {
        try await get(
            path: "api/environments/preview",
            query: ["environmentId": environmentId]
        )
    }

    public func registerEnvironment(id: String, sourceName: String, metadata: [String: JSONValue]) async throws {
        _ = try await postJSON(
            path: "api/environments/register",
            payload: .object([
                "id": .string(id),
                "sourceName": .string(sourceName),
                "metadata": .object(metadata),
            ])
        )
    }

    /// Read-only: ask which `loc:` environments are likely at the given location.
    /// Returns candidates without registering or entering anything.
    public func identifyEnvironments(_ request: IdentifyAvailableRequest) async throws -> [EnvironmentCandidate] {
        try await postLocation(path: "api/environments/identify", request)
    }

    /// Committing: identify, then register/auto-enter the dwell set into the
    /// SessionRoom/agent. Returns the same candidate list.
    public func registerLocation(_ request: IdentifyAvailableRequest) async throws -> [EnvironmentCandidate] {
        try await postLocation(path: "api/environments/register-location", request)
    }

    private func postLocation(path: String, _ request: IdentifyAvailableRequest) async throws -> [EnvironmentCandidate] {
        struct IdentifyResponse: Decodable {
            let candidates: [EnvironmentCandidate]
        }
        var urlRequest = authorizedRequest(url: requestURL(path: path, query: [:]))
        urlRequest.httpMethod = "POST"
        urlRequest.setValue("application/json", forHTTPHeaderField: "Content-Type")
        urlRequest.httpBody = try JSONEncoder().encode(request)
        let (data, response) = try await URLSession.shared.data(for: urlRequest)
        try throwIfErrorResponse(data: data, response: response)
        return try JSONDecoder().decode(IdentifyResponse.self, from: data).candidates
    }

    public func decideEnvironment(environmentId: String, bundleHash: String, decision: String, sessionId: String) async throws {
        _ = try await postJSON(
            path: "api/environments/decision",
            payload: .object([
                "environmentId": .string(environmentId),
                "bundleHash": .string(bundleHash),
                "decision": .string(decision),
                "sessionId": .string(sessionId),
            ])
        )
    }

    public func enterEnvironment(sessionId: String, environmentId: String) async throws -> [String] {
        struct EnterResponse: Decodable {
            let ok: Bool
            let entered: [String]
        }
        let response: EnterResponse = try await post(
            path: "api/environments/enter",
            payload: .object([
                "sessionId": .string(sessionId),
                "environmentId": .string(environmentId),
            ])
        )
        return response.entered
    }

    public func exitEnvironment(sessionId: String, environmentId: String) async throws -> [String] {
        struct ExitResponse: Decodable {
            let ok: Bool
            let entered: [String]
        }
        let response: ExitResponse = try await post(
            path: "api/environments/exit",
            payload: .object([
                "sessionId": .string(sessionId),
                "environmentId": .string(environmentId),
            ])
        )
        return response.entered
    }

    public func environmentList(sessionId: String) async throws -> [EnvironmentListItem] {
        try await get(path: "api/environments/list", query: ["sessionId": sessionId])
    }

    // MARK: - Transport helpers

    private func requestURL(path: String, query: [String: String]) -> URL {
        var components = URLComponents(
            url: baseURL.appending(path: path),
            resolvingAgainstBaseURL: false
        )!
        if !query.isEmpty {
            components.queryItems = query.map { URLQueryItem(name: $0.key, value: $0.value) }
        }
        return components.url!
    }

    private func get<T: Decodable>(path: String, query: [String: String]) async throws -> T {
        let (data, response) = try await URLSession.shared.data(for: authorizedRequest(url: requestURL(path: path, query: query)))
        try throwIfErrorResponse(data: data, response: response)
        return try JSONDecoder().decode(T.self, from: data)
    }

    private func post<T: Decodable>(path: String, payload: JSONValue) async throws -> T {
        var request = authorizedRequest(url: requestURL(path: path, query: [:]))
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.httpBody = try JSONEncoder().encode(payload)
        let (data, response) = try await URLSession.shared.data(for: request)
        try throwIfErrorResponse(data: data, response: response)
        return try JSONDecoder().decode(T.self, from: data)
    }

    private func getJSON(path: String, query: [String: String]) async throws -> JSONValue {
        try await get(path: path, query: query)
    }

    private func postJSON(path: String, payload: JSONValue) async throws -> JSONValue {
        var request = authorizedRequest(url: requestURL(path: path, query: [:]))
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.httpBody = try JSONEncoder().encode(payload)
        let (data, response) = try await URLSession.shared.data(for: request)
        try throwIfErrorResponse(data: data, response: response)
        return try JSONDecoder().decode(JSONValue.self, from: data)
    }

    private func authorizedRequest(url: URL) -> URLRequest {
        var request = URLRequest(url: url)
        if let authToken {
            request.setValue("Bearer \(authToken)", forHTTPHeaderField: "Authorization")
        }
        return request
    }

    private func throwIfErrorResponse(data: Data, response: URLResponse) throws {
        guard let http = response as? HTTPURLResponse, http.statusCode >= 400 else {
            return
        }
        let body = try? JSONDecoder().decode(JSONValue.self, from: data)
        // Fastify 500s put the real message in "message", user-land errors in "error".
        let message = body?["error"]?.stringValue
            ?? body?["message"]?.stringValue
            ?? "Server error (\(http.statusCode))"
        throw RookAPIError(message: message)
    }
}
